docs/admin/ServerSideKeygen.adoc

```diff
@@ -0,0 +1,164 @@
+= ServerSide Key Generation for EE
+Christina Fu <cfu@redhat.com>
+:toc:
+:imagesdir: images
+
+== Overview
+
+Many newer versions of browsers, including Firefox v69 and up, as well as Chrome, have removed the functionality to generate PKI keys and the support for CRMF for key archival. While CLIs such as CRMFPopClient (see "CRMFPopClient --help") or pki (see "pki client-cert-request --help") could be used as a workaround on the Fedora or RHEL platforms, clients on other platforms have been largely left out.
+
+Server-Side Keygen enrollment has been around for a long time since the introduction of Token Key Management System (TMS), where keys could be generated on KRA instead of locally on smart cards. Starting from PKI v.10.5, we adopt a similar mechanism to resolve the browser keygen deficiency issue.   Keys are generated on the server (KRA to be specific) and then transferred securely back to the client in PKCS#12.
+
+NOTE: It is highly recommended that the Server-Side Keygen mechanism only being employed for encryption certificates.
+
+== Functionality Highlights
+
+* Certificate request keys are generated on KRA (Note: KRA must be installed to work with the CA)
+* The profile default plugin, serverKeygenUserKeyDefaultImpl, provides selection to enable or disable key archival (i.e. the “enableArchival” parameter)
+* Support for both RSA and EC keys
+* Support for both manual (agent) approval and automatic approval (e.g. directory password-based)
+
+== Installation configuration
+
+NOTE: A KRA instance is required in addition to the CA for setting up Server-Side Keygen.
+
+NOTE: In the case when the CA and KRA are sharing a Tomcat instance, there is no need to execute the above step to import the transport certificate.
+
+After installing CA and KRA instances, in the case of stand-alone Tomcat web server instances, you’d need to add the KRA’s transport cert to the CA’s nssdb.
+
+First, shut down CA
+
+[literal]
+Systemctl stop pki-tomcatd@<ca_instance_name>.service
+
+Export KRA’s transport certificate into a file
+
+[literal]
+certutil -d /var/lib/pki/<kra_instance_name>/kra/alias -L -n <transport-cert-nickname> -a > kra.cert
+
+Import KRA’s transport cert into CA’s nssdb, using the nickname specified in CA’s CS.cfg
+
+* List the transport cert nickname:
+
+[literal]
+grep "ca.connector.KRA.transportCertNickname" /var/lib/pki/<ca_instance_name>/ca/conf/CS.cfg
+
+* Import using the nickname listed from above step:
+
+[literal]
+certutil -d /var/lib/pki/instance_name/ca/alias -A -t “,,” -n <transportNickName> -i kra.cert
```

Comment:
This should be `/var/lib/pki/<ca_instance_name>/ca/alias`.

In NSS database paths the subsystem name is actually redundant since we can only have one NSS database per instance, so the paths could be simplified as follows:
* `/var/lib/pki/<ca_instance_name>/alias`
* `/var/lib/pki/<kra_instance_name>/alias`

For CS.cfg paths we still need the subsystem names since an instance can contain multiple subsystems.

Reply:
sure.